runtime/gc_glue_java/ArrayletObjectModel.hpp

```diff
@@ -1017,6 +1041,129 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
 #endif /* J9VM_ENV_DATA64 */
 	}
 
+	/**
+	 * Get preserved array layout for the given indexable object
+	 *
+	 * @param forwardedHeader Forwarded header of dstObjectPtr
+	 * @param dstObjectPtr    Pointer to the indexable object whose size is required
+	 */
+	MMINLINE ArrayLayout
+	getPreservedLayout(MM_ForwardedHeader *forwardedHeader, J9IndexableObject *dstObjectPtr)
+	{
+		/* Same as getPreservedClass(forwardedHeader) we don't have access to objectModel so we manually fetch class from forwardedHeader */
+		J9Class* clazz = getPreservedClass(forwardedHeader);
+
+		/* Trivial check for InlineContiguous. */
+		if (0 != getArraySize(dstObjectPtr)) {
+			return GC_ArrayletObjectModel::InlineContiguous;
+		}
+
+		UDATA dataSizeInBytes = getDataSizeInBytes(clazz, getSizeInElements(dstObjectPtr));
+		return getArrayletLayout(clazz, dataSizeInBytes);
+	}
+
+	/**
+	 * Extract the class pointer from an unforwarded object.
+	 *
+	 * This method will assert if the object has been marked as forwarded.
+	 *
+	 * @param[in] pointer to forwardedHeader the MM_ForwardedHeader instance encapsulating the object
+	 * @return pointer to the J9Class from the object encapsulated by forwardedHeader
+	 * @see MM_ForwardingHeader::isForwardedObject()
+	 */
+	MMINLINE J9Class *
+	getPreservedClass(MM_ForwardedHeader *forwardedHeader)
+	{
+		return (J9Class *)((uintptr_t)(forwardedHeader->getPreservedSlot()) & J9GC_J9OBJECT_CLAZZ_ADDRESS_MASK);
+	}
+
+	/**
+	 * Get the preserved hashcode offset of forwarded header.
+	 *
+	 * @param[in]   Pointer to forwardedHeader the MM_ForwardedHeader instance encapsulating the object
+	 * @param clazz Pointer to an J9Class
+	 */
+	MMINLINE uintptr_t
+	getPreservedHashcodeOffset(MM_ForwardedHeader *forwardedHeader, J9Class *clazz)
```

Comment:
you don't have to pass class, you can also get it via that is now avilable
I know it's already available by the caller, but I'm more for cleaner code here rather
recomputing it is not an issue, since it's done farily rarely (only when a hashed array is moved first time)

Reply:
That makes sense!